source/common/router/rds_impl.cc

```diff
@@ -119,5 +125,45 @@ void RdsRouteConfigProviderImpl::registerInitTarget(Init::Manager& init_manager)
   init_manager.registerTarget(*this);
 }
 
+RouteConfigProviderManagerImpl::RouteConfigProviderManagerImpl(
+    Runtime::Loader& runtime, Event::Dispatcher& dispatcher, Runtime::RandomGenerator& random,
+    const LocalInfo::LocalInfo& local_info, ThreadLocal::SlotAllocator& tls)
+    : runtime_(runtime), dispatcher_(dispatcher), random_(random), local_info_(local_info),
+      tls_(tls) {}
+
+std::vector<Router::RouteConfigProviderSharedPtr>
+RouteConfigProviderManagerImpl::routeConfigProviders() {
+  std::vector<Router::RouteConfigProviderSharedPtr> ret;
+  ret.reserve(route_config_providers_.size());
+  for (const auto& element : route_config_providers_) {
+    ret.push_back(element.second.lock());
```

Comment:
OK, I think I get what's going on now. We're using a weak pointer simply as a means to allow handing out shared pointers later when we lock, otherwise we could be using plain `*` pointers. Seems legit. As below, I'd add an `ASSERT` here and comment to explain why this can never fail.